kernel/src/net/socket/netlink/route/message/attr/mod.rs

```diff
@@ -0,0 +1,231 @@
+
+        Ok(())
+    }
+
+    fn as_any(&self) -> &dyn Any;
+}
+
+pub trait ReadAttrFromUser: Sized + NlAttr {
+    type Payload;
+
+    fn new(payload: Self::Payload) -> Self;
+    fn read_payload_from_user(reader: &mut dyn MultiRead, len: usize) -> Result<Self::Payload>;
+    fn read_from_user(reader: &mut dyn MultiRead, header: &CNlAttrHeader) -> Result<Self> {
+        let payload = {
+            let len = header.len as usize - core::mem::size_of_val(header);
+            Self::read_payload_from_user(reader, len)?
+        };
+
+        let res = Self::new(payload);
+
+        Ok(res)
+    }
+}
+
+macro_rules! define_attribute {
+    // First pattern: The payload is CString
+    ($attr_enum:expr, $attr_name:ident, CString, $max_len:expr) => {
+        #[derive(Debug)]
+        pub struct $attr_name {
+            pub value: CString,
+        }
+
+        impl NlAttr for $attr_name {
+            fn type_(&self) -> u16 {
+                $attr_enum as u16
+            }
+
+            fn payload_as_bytes(&self) -> &[u8] {
+                self.value.as_bytes_with_nul()
+            }
+
+            fn as_any(&self) -> &dyn Any {
+                self
+            }
+        }
+
+        impl ReadAttrFromUser for $attr_name {
+            type Payload = CString;
+
+            fn new(value: Self::Payload) -> Self {
+                Self { value }
+            }
+
+            fn read_payload_from_user(
+                reader: &mut dyn MultiRead,
+                len: usize,
+            ) -> Result<Self::Payload> {
+                let max_len = $max_len.min(len);
+                reader.read_cstring_with_max_len(max_len)
+            }
+        }
+    };
+    // Second pattern: The payload is primitive, i.e., any type that implements Pod trait.
+    ($attr_enum:expr, $attr_name:ident, $payload_type: ty) => {
+        #[derive(Debug)]
+        pub struct $attr_name {
+            pub value: $payload_type,
+        }
+
+        impl NlAttr for $attr_name {
+            fn payload_as_bytes(&self) -> &[u8] {
+                self.value.as_bytes()
+            }
+
+            fn type_(&self) -> u16 {
+                $attr_enum as u16
+            }
+
+            fn as_any(&self) -> &dyn Any {
+                self
+            }
+        }
+
+        impl ReadAttrFromUser for $attr_name {
+            type Payload = $payload_type;
+
+            fn new(value: $payload_type) -> Self {
+                Self { value }
+            }
+
+            fn read_payload_from_user(
+                reader: &mut dyn MultiRead,
+                len: usize,
+            ) -> Result<Self::Payload> {
+                if len != core::mem::size_of::<$payload_type>() {
+                    return_errno_with_message!(Errno::EINVAL, "invalid length");
+                }
+
+                reader.read_val::<$payload_type>()
+            }
+        }
+    }; // TODO: The payload is nested attributes
+}
+
+macro_rules! read_attrs_util {
```

Comment:
Again, avoid using macros whenever possible.

I expect something like this to work:
```rust
enum NlAddrAttr {
    Address([u8; 4]),
    Local([u8; 4]),
    // ...
}

impl NlAddrAttr {
    fn read(reader: &mut Reader) -> Self {
        match read_type(&mut reader) {
            AddrAttrClass::ADDRESS => Self::Address(reader.read_val()),
            // ...
    }
}
```